relay-general/src/protocol/session.rs

```diff
@@ -117,6 +125,113 @@ impl SessionUpdate {
+    /// The user agent, if it is different from the one specified in [`SessionAttributes`].
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub user_agent: Option<String>,
+    /// The number of exited sessions that ocurred.
+    #[serde(default, skip_serializing_if = "is_zero")]
+    pub exited: u32,
+    /// The number of errored sessions that ocurred, not including the abnormal and crashed ones.
+    #[serde(default, skip_serializing_if = "is_zero")]
+    pub errored: u32,
+    /// The number of abnormal sessions that ocurred.
+    #[serde(default, skip_serializing_if = "is_zero")]
+    pub abnormal: u32,
+    /// The number of crashed sessions that ocurred.
+    #[serde(default, skip_serializing_if = "is_zero")]
+    pub crashed: u32,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct SessionAggregates {
+    /// A batch of sessions that were started.
+    #[serde(default)]
+    pub aggregates: Vec<AggregateItem>,
+    /// The shared session event attributes.
+    #[serde(rename = "attrs")]
+    pub attributes: SessionAttributes,
+}
+
+impl SessionAggregates {
+    /// Parses a session batch from JSON.
+    pub fn parse(payload: &[u8]) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice(payload)
+    }
+
+    /// Serializes a session batch back into JSON.
+    pub fn serialize(&self) -> Result<Vec<u8>, serde_json::Error> {
+        serde_json::to_vec(self)
+    }
+
+    /// The total number of sessions in this aggregate.
+    pub fn num_sessions(&self) -> u32 {
+        self.aggregates
+            .iter()
+            .map(|i| i.exited + i.errored + i.abnormal + i.crashed)
+            .sum()
+    }
+
+    /// Creates individual session updates from the aggregates.
+    pub fn into_updates_iter(self) -> impl Iterator<Item = SessionUpdate> {
+        let attributes = self.attributes;
+        let mut items = self.aggregates;
+        let mut item_opt = items.pop();
+        std::iter::from_fn(move || loop {
+            let item = item_opt.as_mut()?;
+
+            let (status, errors) = if item.exited > 0 {
+                item.exited -= 1;
+                (SessionStatus::Exited, 0)
+            } else if item.errored > 0 {
+                item.errored -= 1;
+                // when exploding, we create "legacy" session updates that have no `errored` state
+                (SessionStatus::Exited, 1)
+            } else if item.abnormal > 0 {
+                item.abnormal -= 1;
+                (SessionStatus::Abnormal, 1)
+            } else if item.crashed > 0 {
+                item.crashed -= 1;
+                (SessionStatus::Crashed, 1)
+            } else {
+                item_opt = items.pop();
+                continue;
+            };
+            let attributes = if let Some(ua) = item.user_agent.as_ref() {
+                SessionAttributes {
+                    user_agent: Some(ua.clone()),
+                    ..attributes.clone()
```

Comment:
_nit_: Not strictly necessary, but it can be easy to miss new attributes should they be added here. I'd propose to not use the spread operator here, and instead clone individual attributes so that the compiler errors.
This can also be moved into a utility method, either on `AggregateItem` or `SessionAttributes`.